File: android/app/src/main/java/com/ideals/arnav/route/SnapToRoute.kt
package com.ideals.arnav.route

import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

object SnapToRoute {

    /**
     * Snap user position to the nearest point on the route polyline.
     * Uses projection onto line segments in GPS coordinate space.
     */
    fun snap(userLat: Double, userLng: Double, route: List<LatLng>): SnapResult {
        if (route.size < 2) {
            return SnapResult(
                snappedLat = userLat,
                snappedLng = userLng,
                segmentIndex = 0,
                distanceToRoute = 0f,
                distanceAlongRoute = 0f
            )
        }

        var bestDist = Double.MAX_VALUE
        var bestLat = userLat
        var bestLng = userLng
        var bestSegment = 0
        var distAlong = 0.0

        var cumulativeDistance = 0.0

        for (i in 0 until route.size - 1) {
            val a = route[i]
            val b = route[i + 1]

            // Project point onto segment [a, b]
            val projected = projectOntoSegment(
                userLat, userLng,
                a.lat, a.lng,
                b.lat, b.lng
            )

            val dx = userLat - projected[0]
            val dy = userLng - projected[1]
            val dist = dx * dx + dy * dy

            if (dist < bestDist) {
                bestDist = dist
                bestLat = projected[0]
                bestLng = projected[1]
                bestSegment = i

                // Distance along route to the snapped point
                val segStartToSnap = sqrt(
                    (projected[0] - a.lat) * (projected[0] - a.lat) +
                    (projected[1] - a.lng) * (projected[1] - a.lng)
                )
                val segLen = sqrt(
                    (b.lat - a.lat) * (b.lat - a.lat) +
                    (b.lng - a.lng) * (b.lng - a.lng)
                )
                val fraction = if (segLen > 0) segStartToSnap / segLen else 0.0
                distAlong = cumulativeDistance + fraction * segLengthMeters(a, b)
            }

            cumulativeDistance += segLengthMeters(a, b)
        }

        // Convert GPS-space distance to approximate meters
        val distToRoute = com.ideals.arnav.geo.CoordinateConverter.distanceBetween(
            userLat, userLng, bestLat, bestLng
        )

        return SnapResult(
            snappedLat = bestLat,
            snappedLng = bestLng,
            segmentIndex = bestSegment,
            distanceToRoute = distToRoute,
            distanceAlongRoute = distAlong.toFloat()
        )
    }

    /**
     * Project point P onto line segment AB, clamped to [0, 1].
     * Returns [lat, lng] of the projected point.
     */
    private fun projectOntoSegment(
        pLat: Double, pLng: Double,
        aLat: Double, aLng: Double,
        bLat: Double, bLng: Double
    ): DoubleArray {
        val abLat = bLat - aLat
        val abLng = bLng - aLng
        val apLat = pLat - aLat
        val apLng = pLng - aLng

        val ab2 = abLat * abLat + abLng * abLng
        if (ab2 < 1e-20) {
            return doubleArrayOf(aLat, aLng)
        }

        val t = max(0.0, min(1.0, (apLat * abLat + apLng * abLng) / ab2))
        return doubleArrayOf(aLat + t * abLat, aLng + t * abLng)
    }

    private fun segLengthMeters(a: LatLng, b: LatLng): Double {
        return com.ideals.arnav.geo.CoordinateConverter.distanceBetween(
            a.lat, a.lng, b.lat, b.lng
        ).toDouble()
    }
}

File: ios/ARNavigation/ARNavigation/Route/SnapToRoute.swift
import Foundation

/// Snaps user position to the nearest point on a route polyline.
enum SnapToRoute {
    static func snap(
        userLat: Double, userLng: Double,
        route: [LatLng]
    ) -> SnapResult {
        guard route.count >= 2 else {
            return SnapResult(
                snappedLat: userLat, snappedLng: userLng,
                segmentIndex: 0, distanceToRoute: 0, distanceAlongRoute: 0
            )
        }

        var bestDist = Double.greatestFiniteMagnitude
        var bestLat = userLat
        var bestLng = userLng
        var bestSegment = 0
        var distAlong: Double = 0
        var cumulativeDistance: Double = 0

        for i in 0..<(route.count - 1) {
            let a = route[i]
            let b = route[i + 1]

            let projected = projectOntoSegment(
                pLat: userLat, pLng: userLng,
                aLat: a.lat, aLng: a.lng,
                bLat: b.lat, bLng: b.lng
            )

            let dx = userLat - projected.0
            let dy = userLng - projected.1
            let dist = dx * dx + dy * dy

            if dist < bestDist {
                bestDist = dist
                bestLat = projected.0
                bestLng = projected.1
                bestSegment = i

                let segStartToSnap = sqrt(
                    (projected.0 - a.lat) * (projected.0 - a.lat) +
                    (projected.1 - a.lng) * (projected.1 - a.lng)
                )
                let segLen = sqrt(
                    (b.lat - a.lat) * (b.lat - a.lat) +
                    (b.lng - a.lng) * (b.lng - a.lng)
                )
                let fraction = segLen > 0 ? segStartToSnap / segLen : 0
                let segMeters = Double(CoordinateConverter.shared.distanceBetween(
                    lat1: a.lat, lng1: a.lng, lat2: b.lat, lng2: b.lng
                ))
                distAlong = cumulativeDistance + fraction * segMeters
            }

            cumulativeDistance += Double(CoordinateConverter.shared.distanceBetween(
                lat1: a.lat, lng1: a.lng, lat2: b.lat, lng2: b.lng
            ))
        }

        let distToRoute = CoordinateConverter.shared.distanceBetween(
            lat1: userLat, lng1: userLng, lat2: bestLat, lng2: bestLng
        )

        return SnapResult(
            snappedLat: bestLat,
            snappedLng: bestLng,
            segmentIndex: bestSegment,
            distanceToRoute: distToRoute,
            distanceAlongRoute: Float(distAlong)
        )
    }

    /// Project point P onto segment AB, clamped to [0, 1].
    private static func projectOntoSegment(
        pLat: Double, pLng: Double,
        aLat: Double, aLng: Double,
        bLat: Double, bLng: Double
    ) -> (Double, Double) {
        let abLat = bLat - aLat
        let abLng = bLng - aLng
        let apLat = pLat - aLat
        let apLng = pLng - aLng

        let ab2 = abLat * abLat + abLng * abLng
        guard ab2 > 1e-20 else { return (aLat, aLng) }

        let t = max(0, min(1, (apLat * abLat + apLng * abLng) / ab2))
        return (aLat + t * abLat, aLng + t * abLng)
    }
}
